File: internal/ui/commands.go
package ui

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"tui-testing/internal/settings"
	"tui-testing/internal/theme"

	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"
)

// commandSpec is a slash command's entry in the registry that both the
// dispatcher (runCommand) and the inline suggestion dropdown draw from, so
// the two can't drift out of sync.
type commandSpec struct {
	Name string
	Desc string
}

var commandSpecs = []commandSpec{
	{Name: "new", Desc: "Start a new session"},
	{Name: "sessions", Desc: "Switch to a past session"},
	{Name: "theme", Desc: "Change the color theme"},
	{Name: "settings", Desc: "Adjust settings"},
	{Name: "key", Desc: "Set a provider API key"},
	{Name: "agents", Desc: "Configure agent provider/model"},
	{Name: "loader", Desc: "Choose the \"working\" animation"},
	{Name: "interrupt", Desc: "Stop the current response, optionally with a new prompt"},
	{Name: "reload-agents", Desc: "Reload agents/tools/MCP servers from disk"},
	{Name: "exit", Desc: "Quit the app"},
}

// interruptCommandName is /interrupt's registry name — pulled out as a
// const since runCommand also has to recognize "interrupt <prompt>"
// as its own shape, not just the bare command matchCommands/the switch
// below already handle.
const interruptCommandName = "interrupt"

// commandQuery returns the text after a leading "/" in the input box, and
// whether the input is currently in "typing a command" mode at all.
func (a *App) commandQuery() (string, bool) {
	return strings.CutPrefix(a.input.Value(), "/")
}

// matchCommands returns the commands whose name starts with query, in
// registry order. An empty query matches everything.
func matchCommands(query string) []commandSpec {
	query = strings.ToLower(query)
	var matches []commandSpec
	for _, c := range commandSpecs {
		if strings.HasPrefix(c.Name, query) {
			matches = append(matches, c)
		}
	}
	return matches
}

// renderSuggestions draws the inline "/command" dropdown anchored above
// the input bar — a compact, single-line-per-command list, distinct from
// the full-screen popup a selected command may itself go on to open.
func renderSuggestions(s theme.Styles, matches []commandSpec, selected, width int) string {
	// Inner content width: box width (the full terminal width — matching
	// the viewport below it, not width-2, which left a right-edge gap)
	// minus its own border (2) and left/right padding (2).
	rowWidth := max(width-2-2, 0)

	rows := make([]string, len(matches))
	for i, c := range matches {
		name := "/" + c.Name
		if i == selected {
			left := s.SuggestionSelected.Render(" " + name + " ")
			rows[i] = left + s.SuggestionSelectedDesc.Width(max(rowWidth-lipgloss.Width(left), 0)).Render(" "+c.Desc)
			continue
		}
		left := s.SuggestionItem.Render(" " + name)
		rows[i] = left + s.SuggestionDesc.Width(max(rowWidth-lipgloss.Width(left), 0)).Render("  "+c.Desc)
	}
	return s.Suggestions.Width(width).Render(strings.Join(rows, "\n"))
}

// runCommand dispatches a slash command typed into the input (the leading
// "/" already stripped). Unknown commands surface as a system message
// rather than being silently sent as chat. Returns a tea.Cmd for the one
// command (/sessions) that needs an async backend round-trip before it
// has anything to show; every other command acts synchronously and
// returns nil.
func (a *App) runCommand(name string) tea.Cmd {
	trimmed := strings.TrimSpace(name)
	lower := strings.ToLower(trimmed)

	// /interrupt is the one command that takes a free-text argument (the
	// redirect prompt), so it can't go through the bare-word switch
	// below like every other command — matched here first, against both
	// its bare form and "interrupt <anything>", the argument's own
	// casing left untouched since it's about to become chat text.
	switch {
	case lower == interruptCommandName:
		return a.interruptAndSend("")
	case strings.HasPrefix(lower, interruptCommandName+" "):
		return a.interruptAndSend(strings.TrimSpace(trimmed[len(interruptCommandName)+1:]))
	}

	switch lower {
	case "new":
		a.startNewSession()
	case "sessions":
		return a.openSessionsMenu()
	case "theme":
		a.openThemeMenu()
	case "settings":
		a.openSettingsMenu()
	case "key":
		a.openKeyProviderMenu()
	case "agents":
		a.openAgentsMenu()
	case "loader":
		return a.openLoaderMenu()
	case "reload-agents":
		// Same reloadBackend() /agents already uses after a config edit —
		// its own turnInProgress() guard already handles "can't reload
		// mid-turn" correctly for this manual path, nothing extra needed
		// here. See turn.go's concludeTurn for the other way this fires
		// (a reload_agents tool call, deferred until the turn concludes).
		a.setNotice("Reloading agents...")
		return a.reloadBackend()
	case "exit":
		// A typed, deliberate command — unlike ctrl+c (see handleKey),
		// this never needs the "press again to confirm" safety net, so it
		// quits immediately regardless of what else is going on.
		return tea.Quit
	default:
		a.systemMessage("Unknown command: /" + name + " — try /new, /sessions, /theme, /settings, /key, /agents, /loader, /interrupt, /reload-agents, or /exit.")
	}
	return nil
}

func (a *App) openThemeMenu() {
	// Remembered so Esc can revert a live preview instead of leaving
	// whatever theme the cursor happened to be resting on when dismissed.
	a.themeMenuOrigin = a.themeMgr.Current().Name

	names := a.themeMgr.Names()
	items := make([]paletteItem, len(names))
	for i, name := range names {
		desc := ""
		if name == a.themeMenuOrigin {
			desc = "current"
		}
		items[i] = paletteItem{id: name, title: name, desc: desc}
	}
	a.openMenu(paletteTheme, "Choose theme", items)
}

// openLoaderMenu is /loader's entry point — same live-preview shape as
// /theme (see previewWorkingAnim/cancelMenu's paletteLoader case), but
// also has to actually kick off the tick loop itself: unlike a real
// turn, nothing else would ever call startWorkingAnim while just
// browsing the list. Returns a tea.Cmd (openThemeMenu doesn't need to)
// for exactly that reason.
func (a *App) openLoaderMenu() tea.Cmd {
	a.loaderMenuOrigin = workingAnimNames[a.workingAnim.variant]

	items := make([]paletteItem, workingAnimCount)
	for i, name := range workingAnimNames {
		desc := ""
		if name == a.loaderMenuOrigin {
			desc = "current"
		}
		items[i] = paletteItem{id: name, title: name, desc: desc}
	}
	a.openMenu(paletteLoader, "Choose \"working\" animation", items)
	return a.startWorkingAnim()
}

// previewWorkingAnim swaps the active variant immediately, without
// waiting for the menu selection to be confirmed — same reasoning as
// previewTheme, called on every highlight change while /loader has
// focus (see handlePaletteKey).
func (a *App) previewWorkingAnim(name string) {
	a.workingAnim.variant = parseWorkingAnimVariant(name)
}

// openSettingsMenu is /settings' entry point — a top-level category
// picker rather than one flat list of every toggle, so purely visual/TUI
// settings stay visually and structurally separate from agent/tool
// execution policy (currently just "Tool approval mode", but expected to
// grow — see settings.AgentSettings). Selecting a category pushes back
// here (see confirmMenuSelection's paletteSettings case) so Esc/Cancel
// from either sub-page returns to this picker, not straight to the chat.
func (a *App) openSettingsMenu() {
	items := []paletteItem{
		{id: "tui", title: "TUI Settings", desc: "display, popups, tool output"},
		{id: "agent", title: "Agent Settings", desc: "tool approval policy"},
	}
	a.openMenu(paletteSettings, "Settings", items)
}

// openTUISettingsMenu is /settings → "TUI Settings" — purely
// display/interaction toggles, persisted under settings.json's "ui" key.
func (a *App) openTUISettingsMenu() {
	items := []paletteItem{
		{id: "highlight", title: "Highlight user messages", desc: onOff(a.highlightUser)},
		{id: "stream", title: "Stream replies token-by-token", desc: onOff(a.streamReplies)},
		{id: "hitl", title: "Tool approval style", desc: a.hitlMode.String() + " — select to cycle"},
		{id: "verbose", title: "Verbose tool output", desc: onOff(a.verboseTools)},
		{id: "reasoning", title: "Show reasoning text", desc: onOff(a.showReasoning)},
		{id: "popupWidth", title: "Popup width", desc: fmt.Sprintf("%d — select to edit", a.effectivePopupWidth())},
		{id: "popupHeight", title: "Popup height", desc: fmt.Sprintf("%d — select to edit", a.effectivePopupHeight())},
		{id: "toolPreviewLines", title: "Tool output preview lines", desc: fmt.Sprintf("%d — select to edit", a.effectiveToolPreviewMaxLines())},
	}
	a.openMenu(paletteSettingsTUI, "TUI Settings", items)
}

// openAgentSettingsMenu is /settings → "Agent Settings" — agent/tool
// execution policy, persisted under settings.json's "agent" key. Just
// one row today (permission mode); the category exists ahead of having
// much to put in it so later additions (per-tool policy, execution
// targets, ...) have an obvious, already-separated home.
func (a *App) openAgentSettingsMenu() {
	items := []paletteItem{
		{id: "permission", title: "Tool approval mode", desc: a.permissionMode.String() + " — select to cycle"},
	}
	// The execution-target row only appears when the app was wired with a
	// ConfigureTarget closure (main.go always is; a test/embedding harness
	// may not be) — same nil-disables convention as the /agents rows.
	if a.configureTarget != nil {
		items = append(items, paletteItem{id: "target", title: "Tool execution target", desc: a.targetType + " — select to cycle"})
	}
	a.openMenu(paletteSettingsAgent, "Agent Settings", items)
}

// effectiveToolPreviewMaxLines is verbose tool output's actual line
// cap: the configured override if /settings' "Tool output preview
// lines" has been set, else toolPreviewMaxLinesDefault.
func (a *App) effectiveToolPreviewMaxLines() int {
	if a.toolPreviewMaxLines > 0 {
		return a.toolPreviewMaxLines
	}
	return toolPreviewMaxLinesDefault
}

func onOff(v bool) string {
	if v {
		return "on"
	}
	return "off"
}

// dropPendingMessage reports (and clears) a message held by sendMessage
// while /key was collecting a key it never actually got — called from
// wherever the /key flow can be cancelled (either its provider-picker
// first step or the key-entry second step), so a cancelled /key never
// leaves a message silently stranded to be surprise-sent by some later,
// unrelated successful connection.
func (a *App) dropPendingMessage() {
	if a.pendingMessage == "" {
		return
	}
	a.pendingMessage = ""
	a.status = theme.StatusError
	a.systemMessage("Message not sent — no API key set.")
}

// confirmMenuSelection runs the effect of picking id from whichever menu
// is currently open. For the /theme menu the theme is already live (see
// previewTheme) — confirming just makes it permanent instead of
// reverting. Returns whether this was a terminal action: true for a
// setting toggled, a theme/session picked, or a provider saved — false
// when the selection opened another menu instead (see agentsmenu.go's
// /agents steps, and the popupWidth/popupHeight cases below), meaning
// handlePaletteKey should leave whatever new step this call just opened
// alone. When true, handlePaletteKey runs backOrClose rather than an
// unconditional close — a step reached via pushMenuBack goes back to its
// parent instead of leaving the popup; only a true top-level pick (theme,
// loader, session, a plain settings toggle) actually closes it. The
// tea.Cmd, when non-nil, is a terminal action's own async follow-up
// (saving an agent's provider kicks off a backend reload) — batched
// alongside backOrClose's own result regardless of which one fires.
func (a *App) confirmMenuSelection(id string) (bool, tea.Cmd) {
	switch a.paletteKind {
	case paletteTheme:
		a.setNotice("Theme set to " + id + ".")
	case paletteLoader:
		// Explicit rather than relying solely on previewWorkingAnim: if
		// Enter is pressed on the very first row without ever arrowing,
		// no highlight-change event (and so no preview call) ever fired
		// for it.
		a.workingAnim.variant = parseWorkingAnimVariant(id)
		a.persistSettings()
		a.setNotice("Working animation set to " + id + ".")
	case paletteSettings:
		switch id {
		case "tui":
			a.pushMenuBack(func() tea.Cmd { a.openSettingsMenu(); return nil })
			a.openTUISettingsMenu()
			return false, nil
		case "agent":
			a.pushMenuBack(func() tea.Cmd { a.openSettingsMenu(); return nil })
			a.openAgentSettingsMenu()
			return false, nil
		}
	case paletteSettingsTUI:
		switch id {
		case "popupWidth":
			a.pushMenuBack(func() tea.Cmd { a.openTUISettingsMenu(); return nil })
			a.openPopupSizeInput(textPopupPopupWidth, "Set popup width", a.effectivePopupWidth())
			return false, nil
		case "popupHeight":
			a.pushMenuBack(func() tea.Cmd { a.openTUISettingsMenu(); return nil })
			a.openPopupSizeInput(textPopupPopupHeight, "Set popup height", a.effectivePopupHeight())
			return false, nil
		case "toolPreviewLines":
			a.pushMenuBack(func() tea.Cmd { a.openTUISettingsMenu(); return nil })
			a.openPopupSizeInput(textPopupToolPreviewLines, "Set tool output preview lines", a.effectiveToolPreviewMaxLines())
			return false, nil
		default:
			a.toggleSetting(id)
		}
	case paletteSettingsAgent:
		if id == "target" {
			a.cycleExecutionTarget()
		} else {
			a.toggleSetting(id)
		}
	case paletteSessions:
		return true, a.switchSession(id)
	case paletteConfirmDeleteSession:
		return true, a.confirmDeleteSession(id)
	case paletteKeyProvider:
		a.pushMenuBack(func() tea.Cmd { a.openKeyProviderMenu(); return nil })
		a.openKeyInput(id)
		return false, nil
	case paletteAgents:
		a.pushMenuBack(func() tea.Cmd { a.openAgentsMenu(); return nil })
		a.openAgentDetailMenu(id)
		return false, nil
	case paletteAgentDetail:
		return a.confirmAgentDetailSelection(id)
	case paletteAgentProvider:
		return true, a.saveAgentProvider(id)
	case paletteAgentTools:
		// Stays open, same reasoning as the whole page: a checklist is
		// for toggling several things in one visit, not one pick and
		// close like every other menu here.
		a.toggleAgentTool(id)
		return false, nil
	}
	return true, nil
}

// previewTheme applies a theme immediately, without waiting for the menu
// selection to be confirmed, so arrowing through /theme repaints the whole
// app (including the popup itself) live. Called on every highlight change
// while the /theme menu has focus.
func (a *App) previewTheme(name string) {
	a.themeMgr.Set(name)
	a.applyTheme()
}

// cycleExecutionTarget flips the tool execution target between host and
// ssh, persists it, then re-installs the target from settings and reports
// the outcome. The new type is persisted *before* configureTarget runs
// because that reads the target config back from disk (see
// tools.ConfigureTarget). On an SSH failure it reverts to host so the
// persisted/displayed state matches the local fallback the tools actually
// got, rather than claiming "ssh" while running locally.
func (a *App) cycleExecutionTarget() {
	if a.configureTarget == nil {
		return
	}
	if a.targetType == settings.TargetSSH {
		a.targetType = settings.TargetHost
	} else {
		a.targetType = settings.TargetSSH
	}
	a.persistSettings()

	desc, err := a.configureTarget()
	if err != nil {
		a.status = theme.StatusError
		a.systemMessage("Execution target failed: " + err.Error() + " — staying on host.")
		if a.targetType != settings.TargetHost {
			a.targetType = settings.TargetHost
			a.persistSettings()
		}
	} else {
		a.setNotice("Tools now running on: " + desc)
	}
	a.refreshTranscript()
}

func (a *App) toggleSetting(id string) {
	switch id {
	case "highlight":
		a.highlightUser = !a.highlightUser
	case "stream":
		a.streamReplies = !a.streamReplies
	case "hitl":
		a.hitlMode = a.hitlMode.next()
	case "permission":
		a.permissionMode = a.permissionMode.next()
	case "verbose":
		a.verboseTools = !a.verboseTools
	case "reasoning":
		a.showReasoning = !a.showReasoning
	}
	a.persistSettings()
	a.refreshTranscript()
}

// submitNumericSetting is Enter's handler for every /settings numeric
// text field (textPopupPopupWidth/textPopupPopupHeight/
// textPopupToolPreviewLines) — parses the typed value, clamps it into
// that field's own min/max range, and persists it. A non-numeric entry
// is reported and left unchanged rather than silently falling back to
// something the user didn't type.
func (a *App) submitNumericSetting() tea.Cmd {
	raw := strings.TrimSpace(a.keyInput.Value())
	kind := a.textPopupKind

	n, err := strconv.Atoi(raw)
	if err != nil {
		a.systemMessage("Not a number — setting left unchanged.")
		return a.backOrClose(a.closeMenuCmd)
	}

	// Applied before backOrClose below reopens /settings — that reopen
	// rebuilds its rows from the current value right now, so it has to
	// already be live or the row we just edited would flash the old
	// number for one frame.
	switch kind {
	case textPopupPopupWidth:
		a.popupWidth = clampInt(n, popupWidthMin, popupWidthMax)
		a.setNotice(fmt.Sprintf("Popup width set to %d.", a.popupWidth))
	case textPopupPopupHeight:
		a.popupHeight = clampInt(n, popupHeightMin, popupHeightMax)
		a.setNotice(fmt.Sprintf("Popup height set to %d.", a.popupHeight))
	case textPopupToolPreviewLines:
		a.toolPreviewMaxLines = clampInt(n, toolPreviewMaxLinesMin, toolPreviewMaxLinesMax)
		a.setNotice(fmt.Sprintf("Tool output preview lines set to %d.", a.toolPreviewMaxLines))
		// Unlike popup width/height, this affects transcript rendering
		// directly (verbose tool-output blocks), so re-render now rather
		// than waiting for the next unrelated refresh to pick it up.
		a.refreshTranscript()
	}
	a.persistSettings()
	return a.backOrClose(a.closeMenuCmd)
}

func clampInt(n, lo, hi int) int {
	if n < lo {
		return lo
	}
	if n > hi {
		return hi
	}
	return n
}

// persistSettings writes the current UI toggles to settings.json.
// Best-effort, same reasoning as SaveAPIKey in internal/adk: there's
// nowhere safe to report a write failure from inside the TUI, and the
// worst case is just that a toggle doesn't survive a restart.
func (a *App) persistSettings() {
	s := settings.Load()
	s.UI = settings.UISettings{
		HighlightUser:       a.highlightUser,
		StreamReplies:       a.streamReplies,
		HITLMode:            a.hitlMode.String(),
		VerboseTools:        a.verboseTools,
		WorkingAnim:         workingAnimNames[a.workingAnim.variant],
		HideReasoningText:   !a.showReasoning,
		PopupWidth:          a.popupWidth,
		PopupHeight:         a.popupHeight,
		ToolPreviewMaxLines: a.toolPreviewMaxLines,
	}
	// Mutate the loaded Agent section in place rather than replacing it,
	// so fields the UI doesn't own here — the whole Target block (its SSH
	// host/user/key are edited in settings.toml by hand, not in /settings)
	// — survive a settings write instead of being clobbered.
	s.Agent.PermissionMode = a.permissionMode.String()
	s.Agent.Target.Type = a.targetType
	_ = settings.Save(s)
}

func (a *App) systemMessage(text string) {
	a.messages = append(a.messages, ChatMessage{Role: RoleSystem, Content: text, At: time.Now()})
	a.followTranscript()
}

File: internal/webui/web/js/commands.js
// commands.js — the slash-command registry (mirrors commandSpecs in
// internal/ui/commands.go), the cmd* implementations, and the modal
// confirm/navigation logic they share. Template-facing entry points
// (modalClose, modalConfirm, modalHover, selectPaletteItem, submitKey,
// submitAgentDetail) are exported here and put on window by main.js.

import { A, showNotice, loadStatus, saveSettings } from './store.js';
import { applyTheme } from './theme.js';
import { setLoader, startSpinning, LOADER_NAMES } from './anim.js';
import { interrupt } from './stream.js';
import { refreshFileTree } from './filetree.js';
import { onOff, fmtDate, shortId, scrollBottom } from './format.js';
import * as api from './api.js';

export const COMMANDS = [
  { name: '/new',           desc: 'Start a new session' },
  { name: '/sessions',      desc: 'Switch to a past session' },
  { name: '/theme',         desc: 'Change the color theme' },
  { name: '/settings',      desc: 'Adjust settings' },
  { name: '/key',           desc: 'Set a provider API key' },
  { name: '/agents',        desc: 'Configure agent provider/model' },
  { name: '/loader',        desc: 'Choose the "working" animation' },
  { name: '/interrupt',     desc: 'Stop the current response' },
  { name: '/reload-agents', desc: 'Reload agents/tools/MCP servers from disk' },
  { name: '/exit',          desc: 'Quit the app' },
];

export function runCommand(name) {
  switch (name) {
    case '/new':           cmdNew();          break;
    case '/sessions':      cmdSessions();     break;
    case '/theme':         cmdTheme();        break;
    case '/settings':      cmdSettings();     break;
    case '/key':           cmdKey();          break;
    case '/agents':        cmdAgents();       break;
    case '/loader':        cmdLoader();       break;
    case '/interrupt':     interrupt();       break;
    case '/reload-agents': cmdReloadAgents(); break;
    case '/exit':          window.close();    break;
    default: A().sysMsg('Unknown command: ' + name); break;
  }
}

async function cmdNew() {
  try {
    const d = await api.newSession();
    A().sessionId  = d.sessionId;
    A().sessionTag = shortId(d.sessionId);
    A().clearMessages();
    showNotice('Started a new session.');
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

async function cmdSessions() {
  try {
    const list = await api.listSessions();
    if (!list?.length) { showNotice('No past sessions yet.'); return; }
    A().openModal({
      kind:  'sessions',
      title: 'Choose session',
      items: list.map(s => ({
        id:  s.ID,
        name: shortId(s.ID),
        tag:  s.ID === A().sessionId ? 'current' : fmtDate(s.UpdatedAt),
      })),
      idx: Math.max(0, list.findIndex(s => s.ID === A().sessionId)),
    });
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

function cmdTheme() {
  if (!A().themes.length) { A().sysMsg('No themes available.'); return; }
  const origin = A().currentTheme;
  A().openModal({
    kind:  'theme',
    title: 'Choose theme',
    items: A().themes.map(t => ({ id: t.name, name: t.name, tag: t.name === origin ? 'current' : '' })),
    idx:   Math.max(0, A().themes.findIndex(t => t.name === origin)),
    data:  { origin },
  });
}

function cmdSettings() {
  A().openModal({
    kind:  'settings',
    title: 'Settings',
    items: [
      { id: 'tui',   name: 'TUI Settings',  tag: 'display, tool output' },
      { id: 'agent', name: 'Agent Settings', tag: 'tool approval policy' },
    ],
  });
}

function buildTUIItems() {
  const s = A().settings;
  return [
    { id: 'highlight', name: 'Highlight user messages',      tag: onOff(A().highlightUser) },
    { id: 'stream',    name: 'Stream replies token-by-token', tag: onOff(s?.UI?.StreamReplies ?? true) },
    { id: 'verbose',   name: 'Verbose tool output',          tag: onOff(A().verboseTools) },
    { id: 'reasoning', name: 'Show reasoning text',          tag: onOff(A().showReasoning) },
  ];
}

function openTUISettings() {
  A().openModal({ kind: 'settings-tui', title: 'TUI Settings', items: buildTUIItems(), back: cmdSettings });
}

function cmdLoader() {
  if (!A().settings) { A().sysMsg('Settings not loaded.'); return; }
  const origin = A().settings.UI?.WorkingAnim || "Equalizer";
  A().openModal({
    kind:  'loader',
    title: 'Choose animation',
    items: LOADER_NAMES.map(name => ({ id: name, name: name, tag: name === origin ? 'current' : '' })),
    idx:   Math.max(0, LOADER_NAMES.indexOf(origin)),
    data:  { origin },
  });
  // Live preview while the picker is open, whether or not a turn is
  // running — same as the TUI's workingAnimShouldRun (paletteLoader).
  startSpinning();
}

function openAgentSettings() {
  const items = [
    { id: 'permission', name: 'Tool approval mode', tag: modeTag() }
  ];
  if (A().settings?.Agent?.Target) {
    items.push({ id: 'target', name: 'Tool execution target', tag: targetTag() });
  }
  A().openModal({
    kind:  'settings-agent',
    title: 'Agent Settings',
    items: items,
    back: cmdSettings,
  });
}
function modeTag() { return (A().settings?.Agent?.PermissionMode || 'normal') + ' — select to cycle'; }
function targetTag() {
  const t = A().settings?.Agent?.Target?.Type || 'host';
  const desc = A().targetDescription || 'local host';
  return t + ' (' + desc + ') — select to cycle';
}

function cmdKey() { A().openModal({ kind: 'key', title: 'Set API key', items: [] }); }

async function cmdAgents() {
  try {
    const res = await api.getAgents();
    if (!res.ok) { A().sysMsg('Agents not supported: ' + res.statusText); return; }
    A().agentsData = await res.json();
    A().openModal({
      kind:  'agents',
      title: 'Agents',
      items: (A().agentsData.agents || []).map(ag => ({
        id:   ag.ID,
        name: ag.Name || (ag.IsRoot ? 'Root Agent' : ag.ID),
        tag:  (ag.Provider || 'gemini') + ' / ' + (ag.Model || 'default'),
      })),
    });
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

async function cmdReloadAgents() {
  showNotice('Reloading agents...');
  try {
    const res = await api.putAgent({ id: '', provider: '', model: '', tools: [] });
    if (res.ok) { showNotice('Agents reloaded.'); await loadStatus(); }
    else A().sysMsg('Reload failed: ' + res.statusText);
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

// ══════════════════════════════════════════════════════════════════
//  Modal confirmation logic
// ══════════════════════════════════════════════════════════════════

export function modalClose() { A().closeModal(true); }

export function modalHover(i) {
  if (!A().modal || A().modal.idx === i) return;
  A().modal.idx = i;
  if (A().modal.kind === 'theme')  applyTheme(A().modal.items[i].id, false);
  if (A().modal.kind === 'loader') setLoader(A().modal.items[i].id, false);
}

/**
 * Confirm the currently-selected (or a clicked) modal item.
 * i is optional — if supplied, modal.idx is updated first.
 */
export function modalConfirm(i) {
  if (!A().modal) return;
  if (i !== undefined) A().modal.idx = i;
  const item = A().modal.items[A().modal.idx];
  if (!item && A().modal.kind !== 'key' && A().modal.kind !== 'agent-detail') return;
  doModalConfirm(item);
}

function doModalConfirm(item) {
  const kind = A().modal?.kind;
  switch (kind) {
    case 'theme':
      if (item) {
        A().modal.data = null; // commit — don't revert on close
        applyTheme(item.id, true);
        A().closeModal(false);
      }
      break;

    case 'sessions':
      if (item) switchSession(item.id);
      break;

    case 'settings':
      if (!item) break;
      A().closeModal(false);
      // setTimeout defers the sub-menu open so there's no keydown
      // event bubbling into the freshly-opened modal.
      if (item.id === 'tui')   setTimeout(openTUISettings,   0);
      if (item.id === 'agent') setTimeout(openAgentSettings, 0);
      break;

    case 'settings-tui':
      if (item) { toggleTUISetting(item.id); A().modal.items = buildTUIItems(); }
      break;

    case 'settings-agent':
      if (item) toggleAgentSetting(item.id);
      break;

    case 'loader':
      if (item) {
        A().modal.data = null; // commit
        setLoader(item.id, true);
        A().closeModal(false);
      }
      break;

    case 'agents':
      if (item) openAgentDetail(item.id);
      break;
  }
}

function toggleTUISetting(id) {
  const s = A().settings; if (!s) return;
  switch (id) {
    case 'highlight': A().highlightUser = !A().highlightUser; s.UI.HighlightUser = A().highlightUser; break;
    case 'stream':    s.UI.StreamReplies = !s.UI.StreamReplies; break;
    case 'verbose':   A().verboseTools = !A().verboseTools; s.UI.VerboseTools = A().verboseTools; break;
    case 'reasoning': A().showReasoning = !A().showReasoning; s.UI.HideReasoningText = !A().showReasoning; break;
  }
  saveSettings();
}

function toggleAgentSetting(id) {
  const s = A().settings; if (!s) return;
  if (id === 'permission') {
    s.Agent.PermissionMode = s.Agent.PermissionMode === 'normal' ? 'full-auto' : 'normal';
    A().autoAccept = s.Agent.PermissionMode === 'full-auto';
    saveSettings();
    A().modal.items[0].tag = modeTag();
  } else if (id === 'target') {
    s.Agent.Target.Type = s.Agent.Target.Type === 'host' ? 'ssh' : 'host';
    saveSettingsAndReconfigure();
  }
}

async function saveSettingsAndReconfigure() {
  await saveSettings();
  await loadStatus(); // loadStatus's ConfigureTarget call installs the new target
  refreshFileTree();  // ...so the sidebar re-roots at the new cwd immediately
  if (A().modal?.kind === 'settings-agent') {
    A().modal.items = [
      { id: 'permission', name: 'Tool approval mode', tag: modeTag() },
      { id: 'target', name: 'Tool execution target', tag: targetTag() }
    ];
  }
}

async function switchSession(id) {
  A().closeModal(false);
  if (id === A().sessionId) return;
  A().sessionId  = id;
  A().sessionTag = shortId(id);
  A().clearMessages();
  showNotice('Switched to ' + shortId(id) + ' — loading history...');
  try {
    const entries = await api.getTranscript(id);
    replayTranscript(entries);
    showNotice('Switched to ' + shortId(id) + (entries?.length ? ' — history loaded.' : '.'));
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

function replayTranscript(entries) {
  for (const e of entries) {
    if (e.UserText)    A().pushUserMsg(e.UserText);
    else if (e.Text)   A().appendText(e.Text);
    else if (e.Reasoning) A().appendReasoning(e.Reasoning, e.ReasoningMs || 0);
    else if (e.ToolCall)  A().pushToolCall(e.ToolCall.ID, e.ToolCall.Name, e.ToolCall.Args);
    else if (e.ToolResult) A().updateToolResult(e.ToolResult.ID, e.ToolResult.Result);
  }
  A().closeAgentBubble();
  Alpine.nextTick(scrollBottom);
}

function openAgentDetail(agentId) {
  const agent = (A().agentsData?.agents || []).find(a => a.ID === agentId);
  if (!agent) return;
  A().openModal({
    kind:  'agent-detail',
    title: agent.Name || agentId,
    items: [],
    data:  { agentId, agent, tools: A().agentsData?.tools || [] },
    back:  cmdAgents,
  });
}

export async function submitKey() {
  const provider = document.getElementById('mf-provider')?.value || 'gemini';
  const key      = document.getElementById('mf-key')?.value?.trim() || '';
  if (!key) { A().sysMsg('Key cannot be empty.'); return; }
  try {
    const res = await api.putKey(provider, key);
    if (res.ok) {
      A().closeModal(false);
      showNotice('API key saved for ' + provider + '.');
      await loadStatus();
    } else {
      const d = await res.json().catch(() => ({}));
      A().sysMsg('Key error: ' + (d.error || res.statusText));
    }
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

export async function submitAgentDetail() {
  const id       = A().modal?.data?.agentId;
  const provider = document.getElementById('mf-aprov')?.value  || '';
  const model    = document.getElementById('mf-amodel')?.value || '';
  const tools    = Array.from(document.querySelectorAll('#modal-body input[type=checkbox]:checked')).map(b => b.value);
  try {
    const res = await api.putAgent({ id, provider, model, tools });
    A().closeModal(false);
    if (res.ok) { showNotice('Agent config saved. Backend rebuilding...'); await loadStatus(); }
    else A().sysMsg('Save failed: ' + res.statusText);
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

export async function modalDeleteSession() {
  if (A().modal?.kind !== 'sessions') return;
  const item = A().modal.items[A().modal.idx];
  if (!item || !confirm('Delete session ' + item.name + '?')) return;
  try {
    await api.deleteSession(item.id);
    if (item.id === A().sessionId) await cmdNew();
    else await cmdSessions();
  } catch (e) { A().sysMsg('Error: ' + e.message); }
}

// ══════════════════════════════════════════════════════════════════
//  Command palette selection (clicked item)
// ══════════════════════════════════════════════════════════════════
export function selectPaletteItem(i) {
  const cmd = A().paletteMatches[i]; if (!cmd) return;
  document.getElementById('chat-input').value = '';
  A().closePalette();
  setTimeout(() => runCommand(cmd.name), 0);
}
